File: GMUtilsAndroid/GMUtils/GMUtils/src/main/java/gmutils/firebase/fcm/FCM.java
package gmutils.firebase.fcm;

import android.os.Bundle;
import android.text.TextUtils;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.firebase.messaging.FirebaseMessaging;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONObject;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import gmutils.Notifier;
import gmutils.app.BaseApplication;
import gmutils.backgroundWorkers.BackgroundTask;
import gmutils.json.JsonBuilder;
import gmutils.listeners.ActionCallback2;
import gmutils.listeners.ResultCallback;
import gmutils.listeners.ResultCallback2;
import gmutils.listeners.ResultCallback3;
import gmutils.logger.Logger;
import gmutils.logger.LoggerAbs;
import gmutils.net.SimpleHTTPRequest;
import gmutils.storage.GeneralStorage;


public class FCM implements FCMFunctions {
    public static final String EXTRA_FCM_TIMESTAMP = "fcm.timestamp";
    public static final String EXTRA_GOOGLE_MESSAGE_ID = "google.message_id";
    public static final String EXTRA_FROM = "from";

    static FCM _instance;

    public static FCM instance() {
        if (_instance == null) _instance = new FCM();
        return _instance;
    }

    //----------------------------------------------------------------

    public final FirebaseMessaging firebaseMessaging;
    private ResultCallback<String> onDeviceTokenRefresh;
    private SendFcmMessageParameters sendFcmMessageParameters;
    private LoggerAbs logger;

    private FCM() {
        firebaseMessaging = FirebaseMessaging.getInstance();

        LoggerAbs loggerTmp = logger;
        logger = Logger.d();
        printInstallationHint();
        logger = loggerTmp;
    }

    private void printLog(LoggerAbs.ContentGetter content) {
        if (logger != null) {
            logger.print(content);
        }
    }

    public FCM setLogger(LoggerAbs logger) {
        this.logger = logger;
        return this;
    }

    public FCM printInstallationHint() {
        printLog(() -> "[Fcm.init()] -> " +
                "don\"t forget to use FCM.instance.redirectToPendingScreen(); " +
                "in your home screen"
        );
        printLog(() -> "[Fcm.init()] -> " +
                "don\"t forget to define this " +
                Notifier.DEFAULT_NOTIFICATION_CHANNEL_ID +
                " as " +
                "default notification channel (com.google.firebase.messaging.default_notification_channel_id)"
        );

        printLog(() -> "add this to Manifest File::\n" +
                "before <application />\n" +
                "<uses-permission android:name=\"android.permission.POST_NOTIFICATIONS\"/>\n" +
                "then those inside <application />" +
                "<service\n" +
                "    android:name=\"{packageName}.{FCMNotificationService extends gmutils.firebase.fcm.GmFirebaseMessagingService}\"\n" +
                "    android:enabled=\"true\"\n" +
                "    android:exported=\"false\">\n" +
                "    <intent-filter>\n" +
                "        <action android:name=\"com.google.firebase.MESSAGING_EVENT\" />\n" +
                "    </intent-filter>\n" +
                "</service>\n" +
                "\n" +
                "<!-- [optional] -->\n" +
                "<meta-data\n" +
                "    android:name=\"com.google.firebase.messaging.default_notification_icon\"\n" +
                "    android:resource=\"@drawable/fcm_notification_icon\" />\n" +
                "<meta-data\n" +
                "    android:name=\"com.google.firebase.messaging.default_notification_color\"\n" +
                "    android:resource=\"@color/colorAccent\" />\n" +
                "<meta-data\n" +
                "    android:name=\"com.google.firebase.messaging.default_notification_channel_id\"\n" +
                "    android:value=\"" + Notifier.DEFAULT_NOTIFICATION_CHANNEL_ID + "\" />\n" +
                "\n" +
                "\n------------------------------------------------------------\n" +
                "\n" +
                "also add this class::\n" +
                "class {FCMNotificationService} : gmutils.firebase.fcm.GmFirebaseMessagingService() {\n" +
                "     //TODO don't do anything here" +
                "}"
        );

        return this;
    }

    //----------------------------------------------------------------

    @Override
    public FCMFunctions init(
            @NotNull Class<? extends FcmMessageHandler> fcmMessageHandlerClass,
            @NotNull FcmMessageHandler fcmMessageHandler,
            @Nullable ResultCallback<String> onDeviceTokenRefresh,
            @Nullable SendFcmMessageParameters sendFcmMessageParameters
    ) {
        printLog(() -> "Fcm.init");

        this.onDeviceTokenRefresh = onDeviceTokenRefresh;
        this.sendFcmMessageParameters = sendFcmMessageParameters;

        //--------------------------------------------------------------------------

        GmFirebaseMessagingService.onMessage = fcmMessageHandler;

        GmFirebaseMessagingService.registerBackgroundMessageHandler(fcmMessageHandlerClass);

        if (onDeviceTokenRefresh != null) {
            getDeviceToken((s) -> this.onDeviceTokenRefresh.invoke(s));

            GmFirebaseMessagingService.onNewToken = (context, newToken) -> {
                this.onDeviceTokenRefresh.invoke(newToken);
            };
        }

        return this;
    }

    @Override
    public void getDeviceToken(ResultCallback<String> callback) {
        firebaseMessaging.getToken().addOnCompleteListener(task -> {
            if (task.isSuccessful()) {
                callback.invoke(task.getResult());
            }
        });
    }

    //----------------------------------------------------------------------------

    //region subscribe/unsubscribe TOPICs
    private GeneralStorage __prefs;

    private GeneralStorage _prefs() {
        if (__prefs == null) __prefs = GeneralStorage.getInstance("fcm-topics");
        return __prefs;
    }

    @NotNull
    private List<String> savedTopics() {
        var savedTopics = _prefs().retrieveList("FCM_Topics");
        if (savedTopics == null) savedTopics = new ArrayList<>(0);
        return savedTopics;
    }

    @Override
    public void subscribeToTopics(List<String> topics, ResultCallback<Map<String, Boolean>> callback) {
        if (topics == null || topics.isEmpty()) {
            if (callback != null) callback.invoke(null);
            return;
        }

        var result = new HashMap<String, Boolean>();

        var savedTopics = savedTopics();
        if (savedTopics.size() > 0) {
            for (int i = 0; i < topics.size(); i++) {
                var topic = topics.get(i);
                var idx = topics.indexOf(topic);
                if (idx >= 0) {
                    topics.remove(idx);
                    result.put(topic, true);
                } else {
                    result.put(topic, false);
                }
            }
        }

        if (topics.isEmpty()) {
            if (callback != null) {
                callback.invoke(result);
            }
            return;
        }

        for (String topic : topics) {
            try {
                firebaseMessaging.subscribeToTopic(topic);
                printLog(() -> "FCM: subscribed to \"" + topic + "\"");
                result.put(topic, true);
            } catch (Exception e) {
                printLog(() -> "FCM: failed to subscribe to \"" + topic + "\"");
                result.put(topic, false);
            }
        }

        if (callback != null) callback.invoke(result);

        _prefs().saveToList("FCM_Topics", topics);
    }

    @Override
    public void unsubscribeFromTopics(List<String> topics, ResultCallback<Map<String, Boolean>> callback) {
        if (topics == null) {
            if (callback != null) callback.invoke(null);
            return;
        }

        var result = new HashMap<String, Boolean>();

        for (String topic : topics) {
            try {
                firebaseMessaging.unsubscribeFromTopic(topic);
                printLog(() -> "FCM: unsubscribe from " + topic);
                result.put(topic, true);
            } catch (Exception e) {
                printLog(() -> "FCM: failed to unsubscribe from " + topic);
                result.put(topic, false);
            }
        }

        if (callback != null) callback.invoke(result);
    }

    @Override
    public void unsubscribeFromSavedTopics(List<String> exceptedTopics, ResultCallback<Map<String, Boolean>> callback) {
        var savedTopics = savedTopics();

        if (exceptedTopics != null && exceptedTopics.size() > 0) {
            savedTopics.removeAll(exceptedTopics);
        }

        if (savedTopics.size() > 0) {
            unsubscribeFromTopics(savedTopics, callback);
            for (var t : savedTopics) {
                _prefs().removeFromList("FCM_Topics", t);
            }
        } else {
            if (callback != null) callback.invoke(null);
        }
    }

    //endregion

    //----------------------------------------------------------------------------

    //region Send FCM message

    @Override
    public void sendMessageToSpecificDevice(
            String deviceToken,
            //
            String title,
            String message,
            //
            boolean isDataNotification,
            JSONObject dataPayload,
            //
            String channelId,
            String soundFileName,
            //
            ResultCallback<SendFcmCallbackArgs> callback
    ) {
        _sendMessageTo(
                deviceToken + "",
                null,
                //
                title,
                message,
                //
                isDataNotification,
                dataPayload,
                //
                channelId,
                soundFileName,
                //
                callback
        );
    }

    @Override
    public void sendMessageToTopic(
            String topic,
            //
            String title,
            String message,
            //
            boolean isDataNotification,
            JSONObject dataPayload,
            //
            String channelId,
            String soundFileName,
            //
            ResultCallback<SendFcmCallbackArgs> callback
    ) {
        _sendMessageTo(
                null,
                "" + topic,
                //
                title,
                message,
                //
                isDataNotification,
                dataPayload,
                //
                channelId,
                soundFileName,
                //
                callback
        );
    }

    public static final String PAYLOAD_KEY_NAME = "data";

    /**
     * https://firebase.google.com/docs/cloud-messaging/migrate-v1?hl=en&authuser=0#java
     * https://firebase.google.com/docs/cloud-messaging/send-message?authuser=0
     * https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages?authuser=0
     * <deprecated> https://firebase.google.com/docs/cloud-messaging/http-server-ref
     * <deprecated> https://firebase.google.com/docs/cloud-messaging/send-message?hl=en&authuser=0#send-messages-to-topics-legacy
     * FIVE TOPICS IN ONE REQUEST
     */
    private void _sendMessageTo(
            String fcmToken,
            String topic,
            //
            String title,
            String message,
            //
            boolean isDataNotification,
            JSONObject dataPayload,
            //
            String channelId,
            String soundFileName,
            //
            ResultCallback<SendFcmCallbackArgs> callback
    ) {
        try {
            Class.forName("com.google.auth.oauth2.GoogleCredentials");
        } catch (Exception e) {
            throw new RuntimeException(
                    "this dependency is required 'com.google.api-client:google-api-client:2.4.0' .... " +
                            "add it to gradle .... " +
                            "also add this: packagingOptions { resources.excludes.add(\"META-INF/*\") } " +
                            "inside: android {..}"
            );
        }

        if (sendFcmMessageParameters == null)
            throw new IllegalArgumentException("sendFcmMessageParameters must not be null .. use FCM.init()");

        //----------------------------------------------

        BackgroundTask.run(() -> {
            var url = "https://fcm.googleapis.com/v1/projects/" + sendFcmMessageParameters.firebaseProjectId + "/messages:send";

            HashMap<String, String> headers = new HashMap<String, String>();
            headers.put("Content-Type", "application/json; UTF-8");
            headers.put("Authorization", "Bearer " + getAccessToken());

            JsonBuilder messageJson = buildRequestData(
                    fcmToken,
                    topic,
                    title,
                    message,
                    isDataNotification,
                    dataPayload,
                    channelId,
                    soundFileName
            );
            var requestBody = messageJson.toString();

            HttpRequest httpRequest = new HttpRequest(url, headers, requestBody);

            printLog(() -> "sendFcmNotification(" + httpRequest + ")");

            return httpRequest;
        }, (httpRequest) -> {
            httpExecuteDelegate.invoke(httpRequest, callback);
        });
    }

    private JsonBuilder buildRequestData(
            String fcmToken,
            String topic,
            //
            String title,
            String message,
            //
            boolean isDataNotification,
            JSONObject dataPayload,
            //
            String channelId,
            String soundFileName
    ) {
        //https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages?authuser=0
        JsonBuilder notificationBody = JsonBuilder.ofJsonObject();
        if (!TextUtils.isEmpty(fcmToken)) {
            notificationBody.add("token", fcmToken);
        }
        //
        else if (!TextUtils.isEmpty(topic)) {
            notificationBody.add("topic", topic);
        } else {
            throw new IllegalArgumentException("topic or token must set");
        }

        if (!isDataNotification) {
            notificationBody.add(
                    "notification",
                    JsonBuilder.ofJsonObject()
                            .add("title", title)
                            .add("body", message)
            );
        }

        notificationBody.add(
                "android",
                JsonBuilder.ofJsonObject()
                        //.addString("collapse_key", "")
                        //.addString("priority", "NORMAL" /*or HIGH*/)
                        .add(
                                "notification",
                                JsonBuilder.ofJsonObject()
                                        .add("title", title)
                                        .add("body", message)
                                        //.add("notification_priority", "PRIORITY_DEFAULT") //PRIORITY_HIGH - PRIORITY_MAX - PRIORITY_LOW - PRIORITY_MIN
                                        .add("channel_id", channelId)
                                        .add("sound", soundFileName)
                                //.add("icon", "stock_ticker_update")
                                //.add("color", "#7e55c3")
                                //.add("tag", "")
                                //.add("click_action", "")
                                //.add("ticker", "")
                                //.add("sticky", true)
                                //.add("default_sound", true)
                                //.add("visibility", "PUBLIC") //PRIVATE - SECRET
                                //.add("notification_count", 1)
                                //.add("image", "http://....")
                                //.add("direct_boot_ok", false);
                        )
        );

        notificationBody.add(
                "apns",
                JsonBuilder.ofJsonObject()
                        /*.addSubObject(
                                "headers",
                                JsonBuilder.ofJsonObject()
                        )*/
                        .add(
                                "payload",
                                JsonBuilder.ofJsonObject()
                                        .add(
                                                "aps",
                                                JsonBuilder.ofJsonObject()
                                                        .add("title", title)
                                                        .add("body", message)
                                        )
                        )
                        .add("sound", soundFileName)
        );

        notificationBody.add(PAYLOAD_KEY_NAME, dataPayload);

        JsonBuilder messageJson = JsonBuilder.ofJsonObject();
        messageJson.add("message", notificationBody);

        return messageJson;
    }

    private String getAccessToken() {
        final String MESSAGING_SCOPE = "https://www.googleapis.com/auth/firebase.messaging";
        final String[] SCOPES = {MESSAGING_SCOPE};

        InputStream serviceAccountFile;
        if (sendFcmMessageParameters.firebaseServiceAccountFileSource == SendFcmMessageParameters.FileSource.RawResources) {
            serviceAccountFile = BaseApplication
                    .current().
                    getResources().
                    openRawResource((Integer) sendFcmMessageParameters.firebaseServiceAccountFileAddress);
        } else if (sendFcmMessageParameters.firebaseServiceAccountFileSource == SendFcmMessageParameters.FileSource.Storage) {
            try {
                serviceAccountFile = new FileInputStream((String) sendFcmMessageParameters.firebaseServiceAccountFileAddress);
            } catch (FileNotFoundException e) {
                throw new RuntimeException(e);
            }
        } else {
            throw new IllegalStateException(sendFcmMessageParameters.firebaseServiceAccountFileSource + " type is not handled");
        }

        GoogleCredentials googleCredentials = null;
        try {
            googleCredentials = GoogleCredentials
                    .fromStream(serviceAccountFile)
                    .createScoped(Arrays.asList(SCOPES));

            googleCredentials.refresh();

            var value = googleCredentials.getAccessToken().getTokenValue();

            return value;
        } catch (Exception e) {
            throw new RuntimeException(e);
        } finally {
            try {
                serviceAccountFile.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public static class HttpRequest {
        public final String url;
        public final Map<String, String> headers;
        public final String body;

        private HttpRequest(String url, Map<String, String> headers, String body) {
            this.url = url;
            this.headers = headers;
            this.body = body;
        }

        @Override
        public String toString() {
            StringBuilder headersStr = new StringBuilder("{");
            if (headers != null) {
                for (String k : headers.keySet()) {
                    if (headersStr.length() > 1) headersStr.append(",");
                    headersStr
                            .append("\n\t")
                            .append(k)
                            .append(": ")
                            .append(headers.get(k));
                }
            }
            headersStr.append("}");

            return "HttpRequest{\n" +
                    "url='" + url + "'\n" +
                    ", headers=" + headersStr + "\n" +
                    ", body='" + body + "'\n" +
                    '}';
        }
    }

    @NotNull
    public ActionCallback2<HttpRequest, ResultCallback<SendFcmCallbackArgs>, Void> httpExecuteDelegate
            = (request, callback) -> {
        SimpleHTTPRequest.post(
                request.url,
                request.headers,
                request.body,
                null,
                (request2, response) -> {
                    //success response: { "name":"projects/{project_id}/messages/{message_id}" }
                    printLog(() -> "sendFcmNotification(RESPONSE::: " + response + ")");

                    callback.invoke(
                            new FCMFunctions.SendFcmCallbackArgs(
                                    200 == response.getCode(),
                                    0 + response.getCode(),
                                    "" + response.getText(),
                                    "" + (response.getException() != null ? response.getException().getMessage() :
                                            response.getError())
                            )
                    );
                }
        );

        return null;
    };

    //endregion

    //------------------------------------------------------------------------------

    private Bundle lastReceivedNotification;

    public Bundle lastReceivedNotification() {
        return lastReceivedNotification;
    }

    @Override
    public void onActivityStarted(Bundle intentExtras) {
        printLog(() -> "Fcm.onActivityStarted -> extras: " + intentExtras);
        if (intentExtras == null) return;

        if (
                intentExtras.get(EXTRA_GOOGLE_MESSAGE_ID) != null ||
                        intentExtras.get(EXTRA_FCM_TIMESTAMP) != null
        ) {
            lastReceivedNotification = intentExtras;
        }
    }

}

File: GMUtilsFlutter/tutorial/lib/zgmutils/data_utils/firebase/fcm.dart
import 'dart:convert';

import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart' as sharedPrefLib;

import '../../../main.dart' as main;
import '../../../zgmutils/data_utils/storages/locale_preference.dart';
import '../../../zgmutils/utils/notifications.dart';
import '../../utils/logs.dart';

///https://firebase.google.com/docs/cli?authuser=0#mac-linux-auto-script
///https://firebase.flutter.dev/docs/messaging/overview/
///https://pub.dev/packages/flutter_local_notifications#scheduled-notifications-and-daylight-saving-time
///
/// add those to android manifest
//         <meta-data
//             android:name="com.google.firebase.messaging.default_notification_channel_id"
//             android:value="@string/default_channel" />
//
//         <meta-data
//             android:name="com.google.firebase.messaging.default_notification_icon"
//             android:resource="@drawable/notif_icon" />
abstract class IFCM {
  Future<String?> get deviceToken;

  ///must user on main screen starts
  void redirectToPendingScreen();

  Future<void> subscribeToTopics(List<String> topics);

  Future<void> unsubscribeFromTopics();

  Future<bool> sendMessageToSpecificDevice({
    required String deviceToken,
    required String notificationId,
    required String title,
    required String message,
    required String payload,
    AndroidNotificationChannelProperties? channel,
    bool dataNotification = false,
  });

  Future<bool> sendMessageToTopic({
    required String topic,
    required String notificationId,
    required String title,
    required String message,
    required String payload,
    AndroidNotificationChannelProperties? channel,
    bool dataNotification = false,
  });

  int showNotificationProperties(
    String title,
    String body, {
    String? payload,
    int? notificationId,
    AndroidNotificationChannelProperties? customChannel,
    DefaultStyleInformation? androidInformationStyle,
  });
}

//------------------------------------------------------------------------------

class FCM extends IFCM {
  // static String _sentNotificationId = '';

  FirebaseMessaging messaging = FirebaseMessaging.instance;
  FCMConfigurations? fcmConfigurations;

  INotifications localNotifications = Notifications.instance;

  //private constructor
  static FCM? _instance;

  static FCM get instance {
    _instance ??= FCM._();
    return _instance!;
  }

  FCM._();

  Future<void> init(FCMConfigurations? fcmConfigurations) async {
    Logs.print(() => 'Fcm.init');

    this.fcmConfigurations ??= fcmConfigurations;

    //region Local Notification Config
    await localNotifications.init(
      this.fcmConfigurations?.notificationsConfigurations,
    );

    await messaging.setForegroundNotificationPresentationOptions(
      alert: true,
      badge: false,
      sound: true,
    );
    //endregion

    //--------------------------------------------------------------------------

    //region ios permission
    try {
      NotificationSettings settings = await messaging.requestPermission(
        alert: true,
        announcement: false,
        badge: true,
        carPlay: false,
        criticalAlert: false,
        provisional: false,
        sound: true,
      );

      switch (settings.authorizationStatus) {
        case AuthorizationStatus.authorized:
          ////print('FCM: User granted permission');
          break;
        case AuthorizationStatus.provisional:
          ////print('FCM: User granted provisional permission');
          break;
        default:
        ////print('FCM: User declined or has not accepted permission');
      }
    } catch (e) {}
    //endregion

    //--------------------------------------------------------------------------

    //region handle clicking on (notifications)
    FirebaseMessaging.onMessageOpenedApp.listen((RemoteMessage message) {
      Logs.print(() => 'Fcm.FirebaseMessaging.onMessageOpenedApp');
      openCorrespondingScreen(message);
    });
    //endregion

    //region receive messages (notifications)
    //foreground message handler
    FirebaseMessaging.onMessage.listen((RemoteMessage message) {
      _popupNotification(message);
    });

    //background message handler
    FirebaseMessaging.onBackgroundMessage(_handleBackgroundMessage);
    //endregion

    if (fcmConfigurations?.onDeviceTokenRefresh != null) {
      String token;
      try {
        token = await deviceToken ?? '';
      } catch (e) {
        token = '';
      }

      if (token.isNotEmpty) {
        this.fcmConfigurations?.onDeviceTokenRefresh?.call(token);
      }

      FirebaseMessaging.instance.onTokenRefresh.listen((newToken) {
        this.fcmConfigurations?.onDeviceTokenRefresh?.call(newToken);
      });
    }

    Logs.print(() => '[Fcm.init()] -> '
        'don\'t forget to use FCM.instance.redirectToPendingScreen(); '
        'in your home screen');
    Logs.print(() => '[Fcm.init()] -> '
        'don\'t forget to define this "${Notifications.defaultNotificationChannelId}" as '
        'default notification channel (com.google.firebase.messaging.default_notification_channel_id)');
  }

  @override
  Future<String?> get deviceToken async {
    try {
      return await FirebaseMessaging.instance.getToken();
    } catch (e) {
      return null;
    }
  }

  //region handle FCM message
  void openCorrespondingScreen(RemoteMessage message) {
    Logs.print(() => 'Fcm.openCorrespondingScreen');
    openCorrespondingScreenByNotificationJson(message.data);
  }

  void openCorrespondingScreenByNotificationJson(dynamic payload) {
    Logs.print(() => 'Fcm.openCorrespondingScreenByNotificationJson');
    localNotifications.openCorrespondingScreenByNotificationJson(payload);
  }

  void _popupNotification(RemoteMessage message) {
    try {
      LocalePreference().isEn().then((en) {
        _popupNotification2(message, en);
      });
    } catch (e) {
      _popupNotification2(message, null);
    }
  }

  void _popupNotification2(RemoteMessage message, bool? en) {
    Logs.print(
      () => ['FCM._popupNotification2(message: ${message.toMap()}, en: $en)'],
    );

    var title = '${message.notification?.title ?? 'Notification'}•';
    var body = message.notification?.body ?? '';

    var localNotification = resolveNotification(message, en);
    var payload = localNotification.payload;
    if (payload == null) {
      try {
        payload = jsonEncode(message.data);
      } catch (e) {
        Logs.print(() => ['FCM._popupNotification2 ->', e]);
      }
    }

    if (localNotification.allowPopup != false) {
      showNotificationProperties(
        localNotification.title ?? title,
        localNotification.body ?? body,
        payload: payload,
        notificationId: localNotification.notificationId,
        customChannel: localNotification.customChannel,
        androidInformationStyle: localNotification.androidInformationStyle,
      );
    }
  }

  //endregion

  ///must user on main screen starts
  @override
  void redirectToPendingScreen() async {
    Logs.print(() => 'Fcm.redirectToPendingScreen');

    //gets the notification that pushed by Firebase engine
    var msg = await messaging.getInitialMessage();
    if (msg != null) {
      Logs.print(
          () => 'Fcm.redirectToPendingScreen -> messaging.getInitialMessage -> '
              'title: ${msg.notification?.title}, '
              'body: ${msg.notification?.body}, '
              'data: ${msg.data}');

      openCorrespondingScreen(msg);
    } else {
      await localNotifications.redirectToPendingScreen();
    }
  }

  //----------------------------------------------------------------------------

  //region subscribe/unsubscribe TOPICs
  sharedPrefLib.SharedPreferences? __prefs;

  Future<sharedPrefLib.SharedPreferences> get _prefs async {
    __prefs ??= await sharedPrefLib.SharedPreferences.getInstance();
    return __prefs!;
  }

  @override
  Future<void> subscribeToTopics(List<String> topics) async {
    await unsubscribeFromTopics();

    for (var topic in topics) {
      try {
        await messaging.subscribeToTopic(topic);
        Logs.print(() => "FCM: subscribed to topic: \"$topic\"");
      } catch (e) {
        Logs.print(() => "FCM: failed to subscribe to topic: \"$topic\"");
      }
    }

    (await _prefs).setStringList('FCM_Topics', topics);
  }

  @override
  Future<void> unsubscribeFromTopics() async {
    try {
      var topics = (await _prefs).getStringList('FCM_Topics');
      for (var topic in topics ?? []) {
        try {
          await messaging.unsubscribeFromTopic(topic);
          Logs.print(() => "FCM: unsubscribe from topic: $topic");
        } catch (e) {
          Logs.print(() => "FCM: failed to unsubscribe from topic: $topic");
        }
      }
    } catch (e) {}
  }

  //endregion

  //----------------------------------------------------------------------------

  //region Send FCM message
  final IOS_PAYLOAD_KEY_NAME = 'payload';

  @override
  Future<bool> sendMessageToSpecificDevice({
    required String deviceToken,
    required String notificationId,
    required String title,
    required String message,
    required String payload,
    AndroidNotificationChannelProperties? channel,
    bool dataNotification = false,
  }) async {
    return _sendMessageTo(
      to: deviceToken,
      notificationId: notificationId,
      title: title,
      message: message,
      payload: payload,
    );
  }

  @override
  Future<bool> sendMessageToTopic({
    required String topic,
    required String notificationId,
    required String title,
    required String message,
    required String payload,
    AndroidNotificationChannelProperties? channel,
    bool dataNotification = false,
  }) async {
    return _sendMessageTo(
      to: '/topics/$topic',
      notificationId: notificationId,
      title: title,
      message: message,
      payload: payload,
    );
  }

  Future<bool> _sendMessageTo({
    required String to,
    required String notificationId,
    required String title,
    required String message,
    required String payload,
    AndroidNotificationChannelProperties? channel,
    bool dataNotification = false,
  }) async {
    // _sentNotificationId = notificationId;

    //https://firebase.google.com/docs/cloud-messaging/http-server-ref
    //https://firebase.google.com/docs/cloud-messaging/send-message?hl=en&authuser=0#send-messages-to-topics-legacy
    //FIVE TOPICS IN ONE REQUEST

    String? androidChannelId =
        channel?.channelId ?? Notifications.defaultNotificationChannelId;
    SoundFile? sound =
        channel?.soundFile ?? Notifications.defaultNotificationChannelSound;

    var android = {
      'title': title,
      'body': message,
    };
    if (sound != null) {
      android['sound'] = sound.name;
    }
    android['android_channel_id'] = androidChannelId;
    android['channel_id'] = androidChannelId;

    var apns = {
      'aps': {
        'title': title,
        'body': message,
      },
      'title': title,
      'body': message,
      IOS_PAYLOAD_KEY_NAME: payload,
    };
    if (sound != null) {
      apns['sound'] = sound.fileNameWithExtension;
    }

    var notificationBody = {
      'to': to,
      'priority': 'high',
      'data': {
        IOS_PAYLOAD_KEY_NAME: payload,
      },
      'apns': apns,
      'android': android
    };

    if (!dataNotification) {
      notificationBody['notification'] = {
        'title': title,
        'body': message,
      };
    }

    var requestBody = json.encode(notificationBody);

    var url = Uri.parse('https://fcm.googleapis.com/fcm/send');
    var headers = {
      'Content-Type': 'application/json; charset=UTF-8',
      'Authorization':
          'key=${fcmConfigurations?.firebaseProjectMessageKeyForSend}',
    };

    var response = await http.post(
      url,
      headers: headers,
      body: requestBody,
    );

    //if (response.statusCode != 200) {}
    //return response; //{ "message_id": 3598509887081198072 }
    return response.statusCode == 200;
  }

  //endregion

  //----------------------------------------------------------------------------

  @override
  int showNotificationProperties(
    String title,
    String body, {
    String? payload,
    int? notificationId,
    AndroidNotificationChannelProperties? customChannel,
    DefaultStyleInformation? androidInformationStyle,
  }) {
    return localNotifications.showNotification(
      title,
      body,
      payload: payload,
      notificationId: notificationId,
      customChannel: customChannel,
      androidInformationStyle: androidInformationStyle,
    );
  }
}

@pragma('vm:entry-point')
Future<void> _handleBackgroundMessage(RemoteMessage message) async {
  Logs.print(() => 'FCM._handleBackgroundMessage('
      'title: ${message.notification?.title}, '
      'body: ${message.notification?.body}, '
      'data: ${message.data}'
      ')');

  await Firebase.initializeApp();
  FCM.instance.init(FCM.instance.fcmConfigurations);
  FCM.instance._popupNotification(message);
}

@pragma('vm:entry-point')
FcmNotificationProperties resolveNotification(RemoteMessage message, bool? en) {
  /*add this method to main file
    @pragma('vm:entry-point')
    FcmNotificationProperties resolveNotification(RemoteMessage message, bool? en) {}
  */
  return main.resolveNotification(message, en);
}

//------------------------------------------------------------------------------

class FCMConfigurations {
  final NotificationsConfigurations notificationsConfigurations;

  ///get from the firebase console (settings) .... ex:: 'AAAAKbiiUMw:APA91...JRP';
  final String? firebaseProjectMessageKeyForSend;
  final void Function(String)? onDeviceTokenRefresh;

  FCMConfigurations({
    required this.notificationsConfigurations,
    required this.firebaseProjectMessageKeyForSend,
    required this.onDeviceTokenRefresh,
  });
}

class FcmNotificationProperties {
  bool allowPopup;
  int? notificationId;
  String? title;
  String? body;
  String? payload;
  AndroidNotificationChannelProperties? customChannel;
  DefaultStyleInformation? androidInformationStyle;

  FcmNotificationProperties({
    this.allowPopup = true,
    this.notificationId,
    required this.title,
    required this.body,
    this.payload,
    this.customChannel,
    this.androidInformationStyle,
  });
}
